src/Orleans/Messaging/SqlMembershipTable.cs

```diff
@@ -21,489 +21,229 @@ MIT License
             // even if I am not the one who created the table, 
             // try to insert an initial table version if it is not already there,
             // so we always have a first table version row, before this silo starts working.
-            if (tryInitTableVersion)
+            if(tryInitTableVersion)
             {
-                await InitTable();
+                var wasCreated = await InitTableAsync();
+                if(wasCreated)
+                {
+                    logger.Info("Created new table version row.");
+                }
             }
         }
 
+
         public Task InitializeGatewayListProvider(ClientConfiguration config, TraceLogger traceLogger)
         {
-            deploymentId = config.DeploymentId;
-            connectionString = config.DataConnectionString;
+            logger.Verbose3("SqlMembershipTable.InitializeGatewayListProvider called.");
+
+            deploymentId = config.DeploymentId;            
             maxStaleness = config.GatewayListRefreshPeriod;
+
+            //TODO: Orleans does not yet provide the type of database used (to, e.g., to load dlls), so SQL Server is assumed.
+            database = RelationalStorageUtilities.CreateGenericStorageInstance(WellKnownRelationalInvariants.SqlServer, config.DataConnectionString);
+
             return TaskDone.Done;
         }
 
+
         public TimeSpan MaxStaleness
         {
             get { return maxStaleness; }
         }
 
+
         public bool IsUpdatable
         {
             get { return true; }
         }
 
+
         public IList<Uri> GetGateways()
         {
-            var result = new List<Uri>();
-
-            // NOTE: This method is only called from client, so blocking operations should be ok.
-
-            using (var conn = new SqlConnection(connectionString))
+            logger.Verbose3("SqlMembershipTable.GetGateways called.");
+            try
             {
-                conn.Open();
-
-                var cmdStr = String.Format(CLIENT_READ_ROW_SELECT, (Int32)SiloStatus.Active);
-                var command = new SqlCommand(cmdStr);
-                command.Parameters.Add(new SqlParameter { ParameterName = "@deploymentid", DbType = DbType.String, Value = deploymentId });
-                command.Connection = conn;
-
-                // Note: This does blocking read, so must not be called from grain code.
-                using (var results = command.ExecuteReader())
-                {
-                    while (results.Read())
-                    {
-                        int proxyPort = 0;
-                        if (!results.GetSqlInt32(1).IsNull)
-                            proxyPort = results.GetInt32(1);
-                        result.Add(new IPEndPoint(IPAddress.Parse(results.GetString(0)), proxyPort).ToGatewayUri());
-                    }
-                }
+                //TODO: Refactor this to async.
+                return database.ActiveGatewaysAsync(deploymentId).Result;
```

Comment:
I'm not sure if there's an issue for this already, but this sticks into one's eye a bit. :) This looks like requiring refactoring some common functionality too.